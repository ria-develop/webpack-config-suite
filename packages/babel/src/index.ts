export * from './JsxExtensions';
export * from './HotModuleReplacementConfig';
export * from './BabelReactConfig';
export * from './BabelConfig';
