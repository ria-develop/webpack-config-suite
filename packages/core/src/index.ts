import {Configuration} from 'webpack';
import {BaseConfig} from './BaseConfig';
import type {OptionsSchema as WebpackCliOptionsSchema} from '../types/webpack-cli';

export * from './BaseConfig';
export * from './AliasAwareConfig';
export * from './EntryConfig';
export * from './ExtensionsAwareConfig';
export * from './JsExtension';
export * from './WebpackSetup';
export * from './ObjectEntryConfig';
export * from './ResolveAwareConfig';
export * from './CacheThreadRuleAwareConfig';
export * from './RuleSetAwareConfig';
export * from './PluginsAwareConfig';
export * from './Utils';

export type {Options as StyleLoaderOptions} from '../types/style-loader';
export type {Options as PostCssLoaderOptions} from '../types/postcss-loader';
export type {PluginOptions as MiniCssExtractPluginOptions} from '../types/mini-css-extract-plugin';
export type {LoaderOptions as ExtractCssLoaderOptions} from '../types/mini-css-extract-loader';
export type {Options as LessLoaderOptions} from '../types/less-loader';
export type {Options as CssLoaderOptions} from '../types/css-loader';
export type {Options as CacheLoaderOptions} from '../types/cache-loader';
export type {Schema as BabelLoaderOptions} from '../types/babel-loader';

export type WebpackCliOptions = {liveReload?: boolean; hot?: boolean; hotOnly?: boolean} & WebpackCliOptionsSchema;

export type What = typeof BaseConfig | ConfigHandler | (typeof BaseConfig | ConfigHandler)[];
export type ConfigHandler = (config: Configuration, env: unknown, argv: WebpackCliOptions) => What[] | void;
