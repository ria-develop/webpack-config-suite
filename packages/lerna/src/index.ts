export * from './LernaTypeScriptResolvePlugins';
export * from './LernaTypeScriptPlugins';
export * from './LernaTypeScriptReactConfig';
