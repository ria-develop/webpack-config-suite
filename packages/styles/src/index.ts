export * from './CssConfig';
export * from './CssExtension';
export * from './CssPlugins';
export * from './LessConfig';
export * from './LessExtension';
