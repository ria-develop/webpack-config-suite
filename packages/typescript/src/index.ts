export * from './TsExtension';
export * from './TsxExtension';
export * from './TypeScriptConfig';
export * from './TypeScriptPlugins';
export * from './TypeScriptReactConfig';
