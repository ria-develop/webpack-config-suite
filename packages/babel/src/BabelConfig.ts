import {
  BabelLoaderOptions as BabelLoaderOptionsFromSchema,
  BaseConfig,
  CacheThreadRule,
  getLoader,
  JsExtension
} from '@webpack-config-suite/core';
import {RuleSetUseItem} from 'webpack';
import {PluginItem} from '@babel/core';
import {Options as BabelPresetEnvOptions} from '@babel/preset-env';

import {TransformOptions} from 'babel__core';

export type BabelLoaderOptions = TransformOptions & BabelLoaderOptionsFromSchema;

export class BabelConfig extends CacheThreadRule {
  static BABEL_PRESET_ENV: '@babel/preset-env' = '@babel/preset-env';

  static BABEL_PRESET_REACT: '@babel/preset-react' = '@babel/preset-react';

  static BABEL_PLUGIN_HOT_LOADER: 'react-hot-loader/babel' = 'react-hot-loader/babel';

  public get composed(): typeof BaseConfig[] {
    return [JsExtension];
  }

  //--------------------------------------------------------------------------------------------------------------------
  // Preset Env
  //--------------------------------------------------------------------------------------------------------------------
  protected get useReact(): boolean {
    return false;
  }

  protected get babelPresetEnvOptions(): BabelPresetEnvOptions | null {
    return null;
  }

  protected get babelPresetEnv(): (string | BabelPresetEnvOptions | null)[] {
    return [BabelConfig.BABEL_PRESET_ENV, this.babelPresetEnvOptions].filter(Boolean);
  }

  //--------------------------------------------------------------------------------------------------------------------
  // Preset React
  //--------------------------------------------------------------------------------------------------------------------

  protected get babelPresetReactOptions(): unknown {
    return null;
  }

  protected get babelPresetReact(): unknown[] {
    return [BabelConfig.BABEL_PRESET_REACT, this.babelPresetReactOptions].filter(Boolean);
  }

  //--------------------------------------------------------------------------------------------------------------------
  // General
  //--------------------------------------------------------------------------------------------------------------------

  protected get test(): RegExp | string {
    return /\.js$/;
  }

  protected get babelPresets(): PluginItem[] | null {
    return [this.babelPresetEnv, this.useReact && this.babelPresetReact].filter(Boolean);
  }

  protected get babelPlugins(): PluginItem[] | null {
    return this.isHot && [BabelConfig.BABEL_PLUGIN_HOT_LOADER];
  }

  protected get babelLoaderOptions(): BabelLoaderOptions {
    return {presets: this.babelPresets, ...(this.babelPlugins && {plugins: this.babelPlugins})};
  }

  protected get babelLoader(): RuleSetUseItem {
    return getLoader('babel-loader', this.babelLoaderOptions);
  }

  protected get use(): RuleSetUseItem[] {
    return super.use.concat([this.babelLoader]);
  }
}
