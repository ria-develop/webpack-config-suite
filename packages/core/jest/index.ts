export {TestCases} from '../../../jest';
import {processConfig, WebpackCliOptions, What} from '../src';
import {Configuration} from 'webpack';

export function testProcessConfig(
  Entity: What,
  config: Configuration,
  env: unknown = undefined,
  argv: WebpackCliOptions = {}
): void {
  const composed = processConfig(Entity, config, env, argv);
  if (composed) {
    composed.forEach((ComposedConfig) => testProcessConfig(ComposedConfig, config, env, argv));
  }
}
