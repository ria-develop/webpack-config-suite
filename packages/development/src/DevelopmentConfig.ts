import {Plugins, WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import HtmlWebpackPlugin from 'html-webpack-plugin';

export class DevelopmentConfig extends Plugins {
  protected get htmlWebpackPlugin(): HtmlWebpackPlugin {
    return new HtmlWebpackPlugin({template: './template.html'});
  }

  protected get plugins(): Configuration['plugins'] {
    return [(this.isHot || this.isLiveReload) && this.htmlWebpackPlugin];
  }

  protected get devtool(): Configuration['devtool'] {
    return this.isDev ? 'eval' : false;
  }

  constructor(config: Configuration, env: unknown, argv: WebpackCliOptions) {
    super(config, env, argv);
    config.devtool = this.devtool;
  }
}
