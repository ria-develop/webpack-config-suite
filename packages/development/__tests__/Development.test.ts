import {WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {DevelopmentConfig} from '../src';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import {testProcessConfig} from '../../../jest';

describe.each([
  ['development', undefined, {}],
  ['development', undefined, {hot: true}],
  ['development', undefined, {liveReload: true}],
  ['production', undefined, {}]
])(
  'given mode: "%s", env: "%j", argv: "%j"',
  (mode: Configuration['mode'], env: unknown, argv: WebpackCliOptions): void => {
    it('given DevelopmentConfig should match snapshot', () => {
      const config: Configuration = {mode};

      testProcessConfig(DevelopmentConfig, config, env, argv);
      expect(config).toMatchSnapshot();
    });
  }
);

describe('given MyDevelopmentConfig should match snapshot', () => {
  it('should should match snapshot', () => {
    const config: Configuration = {mode: 'development'};

    class MyDevelopmentConfig extends DevelopmentConfig {
      protected get htmlWebpackPlugin(): HtmlWebpackPlugin {
        return undefined;
      }
    }

    testProcessConfig(MyDevelopmentConfig, config);
    expect(config).toMatchSnapshot();
  });
});
