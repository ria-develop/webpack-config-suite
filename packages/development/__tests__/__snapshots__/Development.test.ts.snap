// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given MyDevelopmentConfig should match snapshot should should match snapshot 1`] = `
Object {
  "devtool": "eval",
  "mode": "development",
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"hot":true}" given DevelopmentConfig should match snapshot 1`] = `
Object {
  "devtool": "eval",
  "mode": "development",
  "plugins": Array [
    HtmlWebpackPlugin {
      "assetJson": undefined,
      "childCompilerHash": undefined,
      "hash": undefined,
      "options": Object {
        "base": false,
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "body",
        "meta": Object {},
        "minify": "auto",
        "publicPath": "auto",
        "scriptLoading": "blocking",
        "showErrors": true,
        "template": "./template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
      "version": 4,
    },
  ],
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":true}" given DevelopmentConfig should match snapshot 1`] = `
Object {
  "devtool": "eval",
  "mode": "development",
  "plugins": Array [
    HtmlWebpackPlugin {
      "assetJson": undefined,
      "childCompilerHash": undefined,
      "hash": undefined,
      "options": Object {
        "base": false,
        "cache": true,
        "chunks": "all",
        "chunksSortMode": "auto",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": "body",
        "meta": Object {},
        "minify": "auto",
        "publicPath": "auto",
        "scriptLoading": "blocking",
        "showErrors": true,
        "template": "./template.html",
        "templateContent": false,
        "templateParameters": [Function],
        "title": "Webpack App",
        "xhtml": false,
      },
      "version": 4,
    },
  ],
}
`;

exports[`given mode: "development", env: "undefined", argv: "{}" given DevelopmentConfig should match snapshot 1`] = `
Object {
  "devtool": "eval",
  "mode": "development",
}
`;

exports[`given mode: "production", env: "undefined", argv: "{}" given DevelopmentConfig should match snapshot 1`] = `
Object {
  "devtool": false,
  "mode": "production",
}
`;
